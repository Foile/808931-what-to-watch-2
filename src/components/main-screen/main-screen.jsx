import React, {PureComponent} from "react";
import {arrayOf, shape, string, func, bool} from "prop-types";
import MoviesList from "../movies-list/movies-list";
import GenresList from "../genres-list/genres-list";
import MovieCardInfo from "../movie-card-info/movie-card-info";
import PageFooter from "../page-footer/page-footer";
import ShowMore from "../show-more/show-more";

export default class MainScreen extends PureComponent {
  constructor(props) {
    super(props);
  }

  render() {
    const {movies, onChangeGenre, genres, auth, onLoadMore, isLoadMoreVisible, promo} = this.props;
    return (
      <React.Fragment>
        <section className="movie-card">
          <MovieCardInfo movie={promo} auth={auth}></MovieCardInfo>
        </section>
        <div className="page-content">
          <section className="catalog">
            <h2 className="catalog__title visually-hidden">Catalog</h2>
            <GenresList genres = {genres} onGenreClick={onChangeGenre}/>
            <MoviesList movies = {movies} onHeaderClick={()=>({})}/>
            {isLoadMoreVisible ? <ShowMore onClick = {onLoadMore}></ShowMore> : <div></div>}
          </section>
          <PageFooter/>
        </div>
      </React.Fragment>
    );
  }
}

MainScreen.propTypes = {
  promo: shape({
    name: string
  }),
  movies: arrayOf(shape({
    name: string
  })).isRequired,
  onChangeGenre: func.isRequired,
  auth: shape({}),
  genres: arrayOf(string),
  isLoadMoreVisible: bool,
  onLoadMore: func};
