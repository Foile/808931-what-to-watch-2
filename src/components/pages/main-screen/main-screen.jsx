import React, {PureComponent} from "react";
import {arrayOf, shape, string, func, bool, number} from "prop-types";
import MoviesList from "../../movies-list/movies-list";
import GenresList from "../../genres-list/genres-list";
import MovieCardInfo from "../../movie-card-info/movie-card-info";
import PageFooter from "../../page-footer/page-footer";
import ShowMore from "../../show-more/show-more";
import {connect} from "react-redux";
import getLimitFilteredFilms from "../../../selectors/selectors";
import Constants from "../../../const";
import ActionCreator from "../../../reducers/action-creator/action-creator";


class MainScreen extends PureComponent {
  constructor(props) {
    super(props);
  }

  render() {
    const {films, onChangeGenre, genres, auth, loadMore, limit, promo, isAuthorizationRequired} = this.props;
    const isLoadMoreVisible = limit <= films.length;
    return (
      <React.Fragment>
        <section className="movie-card">
          <MovieCardInfo movie={promo} auth={auth} isAuthorizationRequired={isAuthorizationRequired}></MovieCardInfo>
        </section>
        <div className="page-content">
          <section className="catalog">
            <h2 className="catalog__title visually-hidden">Catalog</h2>
            <GenresList genres = {genres} onGenreClick={onChangeGenre}/>
            <MoviesList movies = {films}/>
            {isLoadMoreVisible ? <ShowMore onClick = {loadMore}></ShowMore> : <div></div>}
          </section>
          <PageFooter/>
        </div>
      </React.Fragment>
    );
  }
}

MainScreen.defaultProps = {
  films: [],
  genre: Constants.DEFAULT_GENRE,
  genres: [Constants.DEFAULT_GENRE]
};

const mapStateToProps = (state, ownProps) => {
  const res = Object.assign({}, ownProps, {
    films: getLimitFilteredFilms(state),
    genres: state.data.genres,
    isAuthorizationRequired: state.user.isAuthorizationRequired,
    auth: state.user.auth,
    limit: state.user.limit,
    promo: state.data.promo
  });
  return res;
};

const mapDispatchToProps = (dispatch) => ({
  onChangeGenre: (genre) => {
    dispatch(ActionCreator.changeGenre(genre));
  },
  onGetMovies: (movies) => dispatch(ActionCreator.getMovies(movies)),
  loadMore: () => dispatch(ActionCreator.loadMore(Constants.FILMS_LIMIT_RATE)),
});


MainScreen.propTypes = {
  loadMore: func,
  limit: number,
  isAuthorizationRequired: bool,
  promo: shape({
    name: string
  }),
  films: arrayOf(shape({
    name: string
  })).isRequired,
  onChangeGenre: func.isRequired,
  auth: shape({}),
  genres: arrayOf(string),
  isLoadMoreVisible: bool,
  onLoadMore: func
};

export {MainScreen};
export default connect(mapStateToProps, mapDispatchToProps)(MainScreen);
