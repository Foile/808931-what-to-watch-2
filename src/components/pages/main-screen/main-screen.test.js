import React from "react";
import renderer from "react-test-renderer";
import {MainScreen} from "./main-screen";
import {BrowserRouter} from "react-router-dom";

it(`App correctly renders after relaunch`, () => {
  const movies = [
    {id: 1, name: `Macbeth`, previewImage: ``, previewVideoLink: ``, genre: `Comedy`},
    {id: 2, name: `Aviator`, previewImage: ``, previewVideoLink: ``, genre: `Comedy`}
  ];
  const tree = renderer
  .create(<BrowserRouter>
    <MainScreen movies = {movies} onChangeGenre={jest.fn()} genres={[`All genres`, `Comedy`]} />
  </BrowserRouter>
  )
  .toJSON();
  expect(tree).toMatchSnapshot();
});
