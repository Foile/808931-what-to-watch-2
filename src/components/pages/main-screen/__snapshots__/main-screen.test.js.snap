// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App correctly renders after relaunch 1`] = `
Array [
  <section
    className="movie-card"
  />,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item catalog__genres-item--active"
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            All genres
          </a>
        </li>
        <li
          className="catalog__genres-item "
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Comedy
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      />
      <div />
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
